docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,482 @@
+  - For example: `RuntimeFeature.IsDynamicCodeCompiled` depends on `IsDynamicCodeSupported`. It should return `false` when `IsDynamicCodeSupported` returns `false`, but it may return `false` even if `IsDynamicCodeSupported` is `true`.
+
+We'll say that a "feature switch property" is also necessarily a "feature guard property" for its defining feature, but not all "feature guard properties" are "feature switch properties".
+
+A "feature switch property" may also be a "feature guard property" for a feature _other than_ the defining feature. For example, we introduced the feature switch property `StartupHookProvider.IsSupported`, which is defined by the feature switch named `"System.StartupHookProvider.IsSupported"`, but additionally serves as a guard for code in the implementation that has `RequiresUnreferencedCodeAttribute`. Startup hook support is disabled whenever we trim out unreferenced code, but may alse be disabled independently by the feature switch.
+
+Similarly, one could imagine introducing a feature switch for `IsDynamicCodeCompiled`. 
+
+
+`IsDynamicCodeSupported` is an example of a feature that has an attribute, a feature switch, and a separate feature guard (`IsDynamicCodeCompiled`), but not all features have all of these bits of functionality.
+
+### Warning behavior
+
+Typically, feature switches come with support (via XML substitutions) for treating feature properties and feature guards as constants when publishing with ILLink/ILCompiler. These tools will eliminate guarded branches. This is useful as a code size optimization, and also as a way to prevent producing warnings for features that have attributes designed to produce warnings at callsites:
+
+```csharp
+UseDynamicCode(); // warns
+
+if (RuntimeFeature.IsDynamicCodeSupported)
+  UseDynamicCode(); // OK, no warning
+
+if (RuntimeFeature.IsDynamicCodeCompiled)
+  UseDynamicCode(); // OK, no warning in ILCompiler, but warns in analyzer
+
+[RequiresDynamicCode("Uses dynamically generated code")]
+static void UseDynamicCode() { }
+```
+
+The ILLink Roslyn analyzer has built-in support for treating `IsDynamicCodeSupported` as a guard for `RequiresDynamicCodeAttribute`, but has no other built-in support.
+
+## Goals
+
+- Teach the ILLink Roslyn analyzer to treat `IsDynamicCodeCompiled` as a guard for `RequiresDynamicCodeAttribute`
+- Allow libraries to define their own feature guard properties
+
+  Libraries should be able to introduce their own properties that can act as guards for `RequiresDynamicCodeAttribute`, or for other features that might produce warnings in the analyzer
+
+- Define an attribute-based model for such feature guards
+
+- Take into account how this would interact with an attribute-based model for feature switches
+
+  We will explore what an attribtute-based model for feature switches would look like to ensure that it interacts well with a model for feature guards. It's possible that we would design both in conjunction if they are naturally related.
```

Comment:
I agree that even if we don't actually implement the switch model, it would be great to have the design for it fully fleshed out. These feel very related.